Tasks/Npm/npmtask.ts

```diff
@@ -1,27 +1,206 @@
+        }
+    } else {
+
+        // new task version with auth support
+        var npmrcPath: string = path.join(cwd, '.npmrc');
+        var tempNpmrcPath : string = getTempNpmrcPath();
+        var debugLog: boolean = tl.getVariable('system.debug') && tl.getVariable('system.debug').toLowerCase() === 'true';
+
+        try{
+            if(tl.osType().toLowerCase() === 'windows_nt') {
+                await runNpmAuthHelperAsync(getNpmAuthHelperRunner(npmrcPath, tempNpmrcPath, debugLog));
+            }
+
+            // set required environment variables for npm execution
+            var npmExecOptions = <trm.IExecOptions>{
+                env: extend({}, process.env)
+            };
+            npmExecOptions.env['npm_config_userconfig'] = tempNpmrcPath;
+            if(debugLog) {
+                npmExecOptions.env['npm_config_loglevel'] =  'verbose';
+            }
+
+            await tryRunNpmConfigAsync(getNpmConfigRunner(debugLog), npmExecOptions);
+            var code : number =  await runNpmCommandAsync(npmRunner, npmExecOptions);
+            cleanUpTempNpmrcPath(tempNpmrcPath);
+            tl.setResult(code, tl.loc('NpmReturnCode', code));
+        } catch (err) {
+            cleanUpTempNpmrcPath(tempNpmrcPath);
+            tl.setResult(tl.TaskResult.Failed, tl.loc('NpmFailed', err.message));
+        }
+    }
+}
+
+async function runNpmAuthHelperAsync(npmAuthRunner: trm.ToolRunner) : Promise<number> {
+    try{
+        var execOptions = <trm.IExecOptions>{
+            env: extend({}, process.env)
+        };
+        execOptions.env = addBuildCredProviderEnv(execOptions.env);
+
+        var code : number = await npmAuthRunner.exec(execOptions);
+        tl.debug('Auth helper exitted with code: ' + code);
+        return Q(code);
+    } catch (err) {
+        // ignore any auth failure and try to run the task.
+        tl.debug(tl.loc('NpmAuthFailed', err.message));
+    }
+}
+
+async function tryRunNpmConfigAsync(npmConfigRunner: trm.ToolRunner, execOptions : trm.IExecOptions) {
+    try {
+        var code = await npmConfigRunner.exec(execOptions);
+    } catch (err) {
+        // 'npm config list' comamnd is run only for debugging/diagnostic
+        // purposes only. Failure of this shouldn't be fatal.
+        tl.warning(tl.loc('NpmConfigFailed', err.Message));
+    }
+}
+
+async function runNpmCommandAsync(npmCommandRunner: trm.ToolRunner, execOptions : trm.IExecOptions) : Promise<number> {
+    try{
+        var code: number = await npmCommandRunner.exec(execOptions);
+        tl.debug('Npm command succeeded with code: ' + code);
+        return Q(code);
+    } catch (err) {
+        tl.debug(tl.loc('NpmFailed', err.message));
+        throw err;
+    }
+}
+
+function shouldUseDeprecatedTask() : boolean {
+    return tl.getVariable('USE_DEPRECATED_TASK_VERSION') && tl.getVariable('USE_DEPRECATED_TASK_VERSION').toLowerCase() === 'true';
+}
+
+function addAuthHelperLocationToPath() {
```

Comment:
The credential provider exe is an implementation detail. We should just hand the tool runner the explicit path to the exe.